src/smartpeak/source/ui/ParameterEditorWidget.cpp

```diff
@@ -0,0 +1,166 @@
+    if (!parameter_.getRestrictionsAsString().empty() && parameter_.getValidStrings().empty())
+    {
+      ImGui::Text(parameter_.getRestrictionsAsString().c_str());
+    }
+
+
+    // Test if user input is valid
+    CastValue validity_test_cast;
+    std::string validity_test_string(input_text_field_.data());
+    Utilities::parseString(validity_test_string, validity_test_cast);
+    if ((validity_test_cast.getTag() == CastValue::Type::INT) && (parameter_.getType() == "float"))
+    {
+      // the parser will consider non decimal-annoted number as int, 
+      // which will not be valid if the parameter is float type.
+      // force conversion.
+      validity_test_cast = static_cast<float>(validity_test_cast.i_);
+    }
+    bool valid = parameter_.isValid(validity_test_cast);
+
+    if (!valid) ImGui::PushStyleVar(ImGuiStyleVar_Alpha, ImGui::GetStyle().Alpha * 0.5f);
+    if (ImGui::Button("Set") && valid)
+    {
+      ParameterSet& user_parameters = application_handler_.sequenceHandler_.getSequence().at(0).getRawData().getParameters();
+      Parameter* existing_parameter = user_parameters.findParameter(function_parameter_, parameter_.getName());
+      if (existing_parameter)
+      {
+        existing_parameter->setValueFromString(std::string(input_text_field_.data()), false);
+      }
+      else
+      {
+        parameter_.setValueFromString(std::string(input_text_field_.data()), false);
+        user_parameters.addParameter(function_parameter_, parameter_);
+      }
+      application_handler_.sequenceHandler_.notifyParametersChanged();
+      ImGui::CloseCurrentPopup();
+    }
+    if (!valid) ImGui::PopStyleVar();
+
+    ImGui::SameLine();
+    if (ImGui::Button("Unset"))
+    {
+      ParameterSet& user_parameters = application_handler_.sequenceHandler_.getSequence().at(0).getRawData().getParameters();
+      if (user_parameters.count(function_parameter_) == 1)
+      {
+        user_parameters[function_parameter_].removeParameter(parameter_.getName());
+        application_handler_.sequenceHandler_.notifyParametersChanged();
+      }
+      ImGui::CloseCurrentPopup();
+    }
+
+    ImGui::SameLine();
+    if (ImGui::Button("Cancel"))
+    {
+      ImGui::CloseCurrentPopup();
+    }
+
+    ImGui::EndPopup();
+  }
+
+  void ParameterEditorWidget::setParameter(const std::string& function_parameter, const Parameter& parameter, const std::string& default_value)
+  { 
+    function_parameter_ = function_parameter;
+    std::ostringstream default_value_stream;
+    default_value_stream << "Default value: " << default_value;
+    default_value_ = default_value_stream.str();
+    parameter_ = parameter;
+    std::string value = parameter.getValueAsString();
+    std::fill(input_text_field_.begin(), input_text_field_.end(), 0);
+    if (value.size() < input_text_field_.size())
+    {
+      std::copy(value.begin(), value.end(), input_text_field_.data());
+    } // else we are in trouble, abnormal long value parameter, do not display.
```

Comment:
You have taken great care to ensure that the user parameters are checked and validated.  Even so, I can imagine that there are edge cases or behavior that the user may experience that they might not understand that would be nice to debug quickly by inspecting the log.  Do you see any opportunities in the code here where some logging could be implemented to better help diagnose problems that the user (or developer) may find when change the parameters?